src/ast/recfun_decl_plugin.h

```diff
@@ -0,0 +1,264 @@
+
+            mutable scoped_ptr<util> m_util;
+            def_map                 m_defs; // function->def
+            case_def_map            m_case_defs; // case_pred->def
+            svector<symbol>         m_def_block;
+            unsigned                m_class_id;
+            util & u() const; // build or return util
+
+            ast_manager & m() { return *m_manager; }
+        public:
+            plugin();
+            virtual ~plugin() override;
+            virtual void finalize() override;
+
+            virtual bool is_fully_interp(sort * s) const override { return false; } // might depend on unin sorts
+        
+            virtual decl_plugin * mk_fresh() override { return alloc(plugin); }
+        
+            virtual sort * mk_sort(decl_kind k, unsigned num_parameters, parameter const * parameters) override { UNREACHABLE(); }
+        
+            virtual func_decl * mk_func_decl(decl_kind k, unsigned num_parameters, parameter const * parameters, 
+                                             unsigned arity, sort * const * domain, sort * range) override;
+                
+            // to start a block of mutually recursive functions
+            void begin_def_block() { m_class_id++; m_def_block.reset(); }
+            void end_def_block();
+
+            def* mk_decl(symbol const& name, unsigned n, sort * params, sort * range);
+
+            def* mk_def(symbol const& name, unsigned n, sort * params, sort * range, var * vars, unsigned n_vars, expr * rhs);
+
+            bool has_def(const symbol& s) const { return m_defs.contains(s); }
+            def const& get_def(const symbol& s) const { return *(m_defs[s]); }
+            def& get_def(symbol const& s) { return *(m_defs[s]); }
+            bool has_case_def(const symbol& s) const { return m_case_defs.contains(s); }
+            case_def& get_case_def(symbol const& s) { return *(m_case_defs[s]); }
+            bool is_declared(symbol const& s) const { return m_defs.contains(s); }
+
+            func_decl * get_decl(def const & d);
+            func_decl * get_decl(case_pred const & d);
+        private:
+            func_decl * mk_fun_pred_decl(unsigned num_parameters, parameter const * parameters, 
+                                         unsigned arity, sort * const * domain, sort * range);
+            func_decl * mk_fun_defined_decl(decl_kind k,
+                                            unsigned num_parameters, parameter const * parameters, 
+                                            unsigned arity, sort * const * domain, sort * range);
+        };
+    }
+
+    // Varus utils for recursive functions
+    class util {
+        friend class decl::plugin;
+        typedef map<symbol, func_decl*, symbol_hash_proc, symbol_eq_proc> func_decl_map;
+        
+        ast_manager &           m_manager;
+        family_id               m_family_id;
+        th_rewriter             m_th_rw;
```

Comment:
so far we haven't used rewriting inside declaration utilities

Reply:
I use it for substituting terms by other terms, at preprocessing time.
For example, `f(x) = p(ite(x>0, a, b))` is compiled into two cases (guarded by `x>0` and `¬(x>0)` respectively), with RHS expression `p(a)` and `p(b)` respectively.